### PR TITLE
Add EIP: Alias ORIGIN to SENDER

### PR DESCRIPTION
**Pull request by cyrusadkisson**
This has been part of the discussion around account abstraction.

### CODE REVIEW

File: EIPS/eip-draft_alias_origin_to_sender.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
+---
+eip: TBD
```
Comment (abcoathup) [inline EIPS/eip-draft_alias_origin_to_sender.md:2]:
```suggestion
eip: 7645
```
Assigning next sequential EIP/ERC/RIP number.
Numbering changed to sequential from 7500.

Please also update the filename.

File: EIPS/eip-7645.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+discussions-to: https://ethereum-magicians.org/t/eip-7645-alias-origin-to-sender/19047
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-03-03
+---
+
+## Abstract
+
+This EIP proposes aliasing the ORIGIN opcode to the SENDER opcode within the Ethereum Virtual Machine (EVM). The purpose of this change is to move Ethereum closer to enabling account abstraction by harmonizing the treatment of externally owned accounts (EOAs) and smart contracts and to address the security concerns associated with the use of ORIGIN that have and will continue to surface in all or most account abstraction proposals.
```
Comment (gaudren) [inline EIPS/eip-7645.md:15]:
```suggestion
This EIP proposes aliasing the `ORIGIN` opcode to the `SENDER` opcode within the Ethereum Virtual Machine (EVM). The purpose of this change is to move Ethereum closer to enabling account abstraction by harmonizing the treatment of externally owned accounts (EOAs) and smart contracts and to address the security concerns associated with the use of `ORIGIN` that have and will continue to surface in all or most account abstraction proposals. 
```
Add backticks to the rest of your proposal too please.

File: EIPS/eip-7645.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+For instance, if an [ERC-4337](./eip-4337.md) bundler has tokens or other authority in a smart contract determined by ORIGIN, any of the transactions it bundles can hijack this authority since ORIGIN remains the bundler address throughout each child transaction.
+
+More apropos in the current context of EVM evolution, the differentiation between the ORIGIN and SENDER opcodes presents a challenge for all account abstraction efforts, such as those outlined in [EIP-7377](./eip-7377.md) and [EIP-3074](./eip-3074.md), because any move towards account abstraction must address the ORIGIN opcode's role, either by modifying or completely bypassing it. Without addressing this, the ORIGIN opcode stands as a barrier to the evolution of Ethereum's account model towards greater flexibility and functionality.
```
Comment (gaudren) [inline EIPS/eip-7645.md:23]: Just a heads up, linking to non-final proposals will prevent yours from advancing.

File: EIPS/eip-7645.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+Definition Change: The ORIGIN opcode (0x32) MUST, in all contexts of execution, return the same value as that returned by the SENDER (also known as CALLER) opcode (0x33).
+
+EVM Implementation: All Ethereum clients MUST implement the following change to the EVM: Whenever the ORIGIN opcode is called, the value to be pushed onto the stack is the current call's sender address, as if the SENDER opcode was executed instead.
+
+Transaction Validation: Transactions MUST be validated as before, with no changes to the transaction structure or processing logic beyond the EVM opcode behavior specified above.
+
+Compatibility: Smart contracts relying on the ORIGIN opcode for obtaining the transaction initiator's address MUST be reviewed to ensure they function correctly under the new definition and worked-around or avoided if this EIP introduces breaking changes.
```
Comment (gaudren) [inline EIPS/eip-7645.md:37]: Please move this section to the Backwards Compatibility section. This standard can't really change behaviour of already deployed contracts.

File: EIPS/eip-7645.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+Implementers are encouraged to provide feedback on this specification and report any potential issues encountered during the implementation or testing phases.
```
Comment (gaudren) [inline EIPS/eip-7645.md:39]: Proposal should be written as if they are already final. Assuming this proposal is accepted, this line won't make sense 10 years from now.

File: EIPS/eip-7645.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+## Rationale
+
+The rationale behind aliasing ORIGIN to SENDER is to:
+
+Facilitate Account Abstraction: Elegantly nullify a universal barrier to account abstraction, enabling more flexible and powerful account models in Ethereum.
+
+Enhance Security: Eliminate the security vulnerabilities associated with differentiating between the original transaction initiator and the immediate caller.
+
+Clean up tech debt and simplify the EVM Model: Reduce the complexity of the EVM's transaction and execution model by removing an outdated and deprecated feature, making future changes easier and safer.
+
```
Comment (gaudren) [inline EIPS/eip-7645.md:50]: This reads as Motivation, not Rationale. The Rationale section should explain technical choices made in the EIP.

File: EIPS/eip-7645.md (hunk 3 above)
Comment (cyrusadkisson) [reply EIPS/eip-7645.md:23]: I'll remove the references to 7377 and 3074, but why is EIP-4337 still "draft" if it's in use today?

**Review by mattstam — COMMENTED** [review]
If all the existing documentation around `tx.origin` is considered, I think it becomes quite clear that this change is not needed. I have yet to see a resource that does not explicitly state `tx.origin` should generally not be used.

We should not assume account abstraction ever goes to 100% usage (and it is far from it now). And since that's the case, `tx.origin` still does have niche use-cases, like checking which address paid for the transaction.

In other words, I think this change would cause more confusion than it solves.

**Comment by SamWilsn** [comment]
@mattstam please leave technical comments like this one on the [discussion thread](https://ethereum-magicians.org/t/eip-7645-alias-origin-to-sender/19047). Pull request comments tend to get lost when the pull request gets merged.

File: EIPS/eip-7645.md (hunk 3 above)
Comment (gaudren) [reply EIPS/eip-7645.md:23]: That's up to the authors of ERC-4337.

**Review by lightclient — APPROVED** [review]

**Pull request merged**